### PR TITLE
Decoupling scheduler creation from creation of scheduler.Config struc…

### PR DESCRIPTION
**Pull request by jayunit100**
**What this PR does / why we need it**:

Adds functionality to the scheduler to initialize from an Configurator interface, rather then via a Config struct.

**Which issue this PR fixes** 

Reduces coupling to `scheduler.Config` data structure format so that we can proliferate more interface driven composition of scheduler components.

### CODE REVIEW

**Comment by jayunit100** [comment]
This is a first pass at making `Scheduler` creation interface driven.  The idea is that, once we have a mechanism for creating from `Configurator`, we can start phasing out creation using `New`.  I didn't refactor the unit test because it should *still* test Config, but in time
- maybe `Config` -> `config`
- and then `config` wont need direct unit tests
- and then we can update the scheduler_test to create using a Configurator.

The long term goal here is to simply make scheduler creation completely interface driven, but removing `Config` isn't a good idea right now, since we need to audit what functionality the scheduler thoroughly, and 1.6 is just a stabilization release.

File: plugin/pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -100,14 +101,15 @@ type Config struct {
 	// question, and the error
 	Error func(*v1.Pod, error)
 
-	// Recorder is the EventRecorder to use
+	// recorder is the EventRecorder to use
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/scheduler.go]: Please revert this change.

File: plugin/pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -78,6 +78,7 @@ type Configurator interface {
 	CreateFromKeys(predicateKeys, priorityKeys sets.String, extenders []algorithm.SchedulerExtender) (*Config, error)
 }
 
+// TODO jayunit100 over time we should make this struct a hidden implementation detail of the scheduler.
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/scheduler.go]: s/jayunit100 //

File: plugin/pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -100,14 +101,15 @@ type Config struct {
...
+	// recorder is the EventRecorder to use
 	Recorder record.EventRecorder
 
 	// Close this to shut down the scheduler.
 	StopEverything chan struct{}
 }
 
-// New returns a new scheduler.
+// New returns a new scheduler.  This can give way over time to the "NewFromConfigurator" function,
+// which is interface driven.
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/scheduler.go]: Shouldn't we add a TODO to remove this function and replace its usage with NewFromConfigurator?

File: plugin/cmd/kube-scheduler/app/server.go [hunk 4]
```diff
@@ -171,33 +174,47 @@ func createClient(s *options.SchedulerServer) (*clientset.Clientset, error) {
...
 		}
-		return configFactory.CreateFromConfig(policy)
-	}
+		return sc.Configurator.CreateFromConfig(policy)
 
-	// if the config file isn't provided, use the specified (or default) provider
-	config, err := configFactory.CreateFromProvider(s.AlgorithmProvider)
-	if err != nil {
-		return nil, err
+	} else {
```
Comment (wojtek-t) [inline plugin/cmd/kube-scheduler/app/server.go]: I would change the order of this if to avoid else. How about:
```
if _, err := os.Stat(sc.policyFile); err != nil {
  return sc.Configurator.CreateFromProvider(sc.algorithmProvider
}
configData, err := ...
...
```
?

**Review by wojtek-t — COMMENTED** [review]
Just some minor comments.

**Comment by jayunit100** [comment]
@k8s-bot gci gke e2e test this.

**Comment by jayunit100** [comment]
@k8s-bot verify unit test this.

**Comment by jayunit100** [comment]
@k8s-bot unit test this

**Comment by jayunit100** [comment]
@wojtek-t ok, fixed a bug in the scheduler_perf part that was breaking unit tests.  should be ready for final review now.  PTAL and thanks !

**Comment by jayunit100** [comment]
NEVERMIND, something is odd here, i think I force pushed to the wrong branch, will update shortly...

**Comment by jayunit100** [comment]
@k8s-bot verify test this

**Comment by jayunit100** [comment]
@k8s-bot verify test this.

File: plugin/cmd/kube-scheduler/app/server.go [hunk 5]
```diff
@@ -78,24 +78,20 @@ func Run(s *options.SchedulerServer) error {
 	if err != nil {
 		return fmt.Errorf("unable to create kube client: %v", err)
 	}
-	config, err := createConfig(s, kubecli)
+	recorder := createRecorder(kubecli, s)
+	sched, err := createScheduler(s, kubecli, recorder)
 	if err != nil {
-		return fmt.Errorf("failed to create scheduler configuration: %v", err)
+		fmt.Errorf("Error creating scheduler: %v", err)
```
Comment (k82cn) [inline plugin/cmd/kube-scheduler/app/server.go]: s/Error/error

File: plugin/cmd/kube-scheduler/app/server.go [hunk 6]
```diff
@@ -171,33 +174,45 @@ func createClient(s *options.SchedulerServer) (*clientset.Clientset, error) {
...
+	if _, err := os.Stat(sc.policyFile); err != nil {
+		return sc.Configurator.CreateFromProvider(sc.algorithmProvider)
 	}
 
-	// if the config file isn't provided, use the specified (or default) provider
-	config, err := configFactory.CreateFromProvider(s.AlgorithmProvider)
+	// policy file is valid, try to create a configuration from it.
+	var (
+		policy     schedulerapi.Policy
+		configData []byte
```
Comment (k82cn) [inline plugin/cmd/kube-scheduler/app/server.go]: configData is not necessary. `ioutil.ReadFile` will create it.

File: plugin/cmd/kube-scheduler/app/server.go [hunk 7]
```diff
@@ -171,33 +174,45 @@ func createClient(s *options.SchedulerServer) (*clientset.Clientset, error) {
...
+	)
+	configData, err := ioutil.ReadFile(sc.policyFile)
```
Comment (k82cn) [inline plugin/cmd/kube-scheduler/app/server.go]: Do we need to consider symlink here?

**Comment by jayunit100** [comment]
@k82cn yeah, i agree with those changes.  No harm including them including them even though that code was just being moved.

**Comment by k82cn** [comment]
@jayunit100 , agree, thanks :).

**Comment by jayunit100** [comment]
fixed the bazel issue i had, PTAL @wojtek-t thanks !

File: plugin/cmd/kube-scheduler/app/server.go [hunk 8]
```diff
@@ -171,33 +174,42 @@ func createClient(s *options.SchedulerServer) (*clientset.Clientset, error) {
...
-	config.Recorder = eventBroadcaster.NewRecorder(v1.EventSource{Component: s.SchedulerName})
-	eventBroadcaster.StartLogging(glog.Infof)
-	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: kubecli.Core().Events("")})
+// createScheduler encapsulates the entire creation of a runnable scheduler.
+func createScheduler(s *options.SchedulerServer, kubecli *clientset.Clientset, recorder record.EventRecorder) (*scheduler.Scheduler, error) {
+	configurator := factory.NewConfigFactory(kubecli, s.SchedulerName, s.HardPodAffinitySymmetricWeight, s.FailureDomains)
 
-	return config, nil
+	// Rebuild the configurator with a default Create(...) method.
+	configurator = &schedulerConfigurator{
```
Comment (jayunit100) [inline plugin/cmd/kube-scheduler/app/server.go]: @sjug fyi

File: plugin/cmd/kube-scheduler/app/server.go [hunk 9]
```diff
@@ -78,24 +78,20 @@ func Run(s *options.SchedulerServer) error {
 	if err != nil {
 		return fmt.Errorf("unable to create kube client: %v", err)
 	}
-	config, err := createConfig(s, kubecli)
+	recorder := createRecorder(kubecli, s)
+	sched, err := createScheduler(s, kubecli, recorder)
 	if err != nil {
-		return fmt.Errorf("failed to create scheduler configuration: %v", err)
+		return fmt.Errorf("Error creating scheduler: %v", err)
```
Comment (wojtek-t) [inline plugin/cmd/kube-scheduler/app/server.go]: s/Error/error/ (convention is that we return errors starting with small letter)

File: plugin/pkg/scheduler/scheduler.go [hunk 10]
```diff
@@ -107,7 +112,7 @@ type Config struct {
 	StopEverything chan struct{}
 }
 
-// New returns a new scheduler.
+// New returns a new scheduler.  TODO replace this with NewFromConfigurator.
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/scheduler.go]: nit: Move TODO to the next line

**Review by wojtek-t — COMMENTED** [review]
Just some super minor comments. Other than that LGTM
